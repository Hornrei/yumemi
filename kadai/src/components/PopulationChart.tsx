// PopulationChart.js
import React from "react";
import PropTypes from "prop-types";
import {
  LineChart,
  Line,
  XAxis,
  YAxis,
  CartesianGrid,
  Tooltip,
  Legend,
  ResponsiveContainer,
} from "recharts";

function PopulationChart({
  data,
  selectedPrefectures,
  colorPalette,
  prefectures,
}) {
  const defaultYear = [
    { year: "1960" },
    { year: "1965" },
    { year: "1970" },
    { year: "1975" },
    { year: "1980" },
    { year: "1985" },
    { year: "1990" },
    { year: "1995" },
    { year: "2000" },
    { year: "2005" },
    { year: "2010" },
    { year: "2015" },
    { year: "2020" },
    { year: "2025" },
    { year: "2030" },
    { year: "2035" },
    { year: "2040" },
    { year: "2045" },
  ];

  const formattedData = data.reduce((acc, { prefCode, data }) => {
    data.forEach((item, index) => {
      if (!acc[index]) {
        acc[index] = { year: item.year };
      }
      acc[index][prefectures[prefCode - 1].prefName] = item.value;
    });
    return acc;
  }, []);

  return (
    <ResponsiveContainer width="100%" height={400}>
      <LineChart
        data={formattedData.length > 0 ? formattedData : defaultYear}
        margin={{ top: 5, right: 30, left: 30, bottom: 5 }}
      >
        <CartesianGrid strokeDasharray="3 3" />
        <XAxis dataKey="year" />
        <YAxis tickFormatter={(value) => value.toLocaleString()} />
        <Tooltip />
        <Legend verticalAlign="top" />
        {formattedData.length > 0 &&
          selectedPrefectures.map((prefCode, index) => {
            const prefName = prefectures[prefCode - 1].prefName;
            return (
              <Line
                key={prefCode}
                type="monotone"
                dataKey={prefName}
                name={prefName}
                stroke={colorPalette[index % colorPalette.length]}
                activeDot={{ r: 8 }}
              />
            );
          })}
      </LineChart>
    </ResponsiveContainer>
  );
}

PopulationChart.propTypes = {
  data: PropTypes.array.isRequired,
  selectedPrefectures: PropTypes.array.isRequired,
  colorPalette: PropTypes.array.isRequired,
  prefectures: PropTypes.array.isRequired,
};

export default PopulationChart;
