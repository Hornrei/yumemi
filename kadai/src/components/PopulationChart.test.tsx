// src/PopulationChart.test.js
import React from "react";
import { render } from "@testing-library/react";
// @ts-expect-error TS(6142): Module './PopulationChart' was resolved to 'C:/Git... Remove this comment to see the full error message
import PopulationChart from "./PopulationChart";

const data = [
  { prefCode: 1, data: [{ year: 1960, value: 5000000 }] },
  { prefCode: 2, data: [{ year: 1960, value: 3000000 }] },
];

const selectedPrefectures = [1, 2];
const colorPalette = ["#8884d8", "#82ca9d"];
const prefectures = [
  { prefCode: 1, prefName: "北海道" },
  { prefCode: 2, prefName: "青森県" },
];

// @ts-expect-error TS(2582): Cannot find name 'test'. Do you need to install ty... Remove this comment to see the full error message
test("renders population chart", () => {
  const { container } = render(
    // @ts-expect-error TS(17004): Cannot use JSX unless the '--jsx' flag is provided... Remove this comment to see the full error message
    <PopulationChart
      data={data}
      selectedPrefectures={selectedPrefectures}
      colorPalette={colorPalette}
      prefectures={prefectures}
    />,
  );
  // @ts-expect-error TS(2304): Cannot find name 'expect'.
  expect(container).toMatchSnapshot();
});
