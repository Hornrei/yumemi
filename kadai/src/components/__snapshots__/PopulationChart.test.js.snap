// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders population chart 1`] = `
<div>
  <div
    class="recharts-responsive-container"
    style="width: 100%; height: 400px; min-width: 0;"
  />
</div>
`;
